File: python-client/p2pshare/crypto/session.py
"""AES-256-GCM session encryption/decryption with sequence number tracking.

Wire format for encrypted payloads:
    nonce (12 bytes) || ciphertext || tag (16 bytes)

GCM provides both confidentiality and integrity.
"""

import secrets
import struct

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from p2pshare.errors import IntegrityError, ProtocolError


NONCE_SIZE = 12
TAG_SIZE = 16
MAX_PAYLOAD_SIZE = 64 * 1024 * 1024  # 64 MB


class SessionCipher:
    """Encrypts/decrypts messages for an established session.

    Tracks send and receive sequence numbers for replay protection.
    """

    def __init__(self, session_key: bytes):
        if len(session_key) != 32:
            raise ValueError("Session key must be 32 bytes")
        self._gcm = AESGCM(session_key)
        self._send_seq = 0
        self._recv_seq = 0

    @property
    def send_seq(self) -> int:
        return self._send_seq

    @property
    def recv_seq(self) -> int:
        return self._recv_seq

    def encrypt(self, plaintext: bytes) -> bytes:
        """Encrypt plaintext. Returns nonce || ciphertext || tag."""
        nonce = secrets.token_bytes(NONCE_SIZE)
        ct = self._gcm.encrypt(nonce, plaintext, None)
        self._send_seq += 1
        return nonce + ct

    def decrypt(self, payload: bytes) -> bytes:
        """Decrypt a payload of nonce || ciphertext || tag.

        Raises IntegrityError if the GCM tag verification fails.
        """
        if len(payload) < NONCE_SIZE + TAG_SIZE:
            raise ProtocolError("Encrypted payload too short")
        nonce = payload[:NONCE_SIZE]
        ct = payload[NONCE_SIZE:]
        try:
            plaintext = self._gcm.decrypt(nonce, ct, None)
        except Exception:
            raise IntegrityError(
                "Message integrity check failed — possible tampering or data corruption."
            )
        self._recv_seq += 1
        return plaintext


def encrypt_raw(key: bytes, plaintext: bytes) -> bytes:
    """One-shot AES-256-GCM encryption. Returns nonce || ciphertext || tag."""
    if len(key) != 32:
        raise ValueError("Key must be 32 bytes")
    nonce = secrets.token_bytes(NONCE_SIZE)
    ct = AESGCM(key).encrypt(nonce, plaintext, None)
    return nonce + ct


def decrypt_raw(key: bytes, payload: bytes) -> bytes:
    """One-shot AES-256-GCM decryption of nonce || ciphertext || tag."""
    if len(key) != 32:
        raise ValueError("Key must be 32 bytes")
    if len(payload) < NONCE_SIZE + TAG_SIZE:
        raise IntegrityError("Encrypted payload too short")
    nonce = payload[:NONCE_SIZE]
    ct = payload[NONCE_SIZE:]
    try:
        return AESGCM(key).decrypt(nonce, ct, None)
    except Exception:
        raise IntegrityError(
            "Message integrity check failed — possible tampering or data corruption."
        )

File: go-client/internal/crypto/session.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"errors"
	"fmt"
	"io"
)

const (
	NonceSize      = 12
	TagSize        = 16
	MaxPayloadSize = 64 * 1024 * 1024 // 64 MB
)

// SessionCipher handles AES-256-GCM encryption/decryption for an established session.
type SessionCipher struct {
	gcm     cipher.AEAD
	SendSeq int
	RecvSeq int
}

// NewSessionCipher creates a new session cipher from a 32-byte session key.
func NewSessionCipher(sessionKey []byte) (*SessionCipher, error) {
	if len(sessionKey) != 32 {
		return nil, errors.New("session key must be 32 bytes")
	}
	block, err := aes.NewCipher(sessionKey)
	if err != nil {
		return nil, fmt.Errorf("aes.NewCipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("cipher.NewGCM: %w", err)
	}
	return &SessionCipher{gcm: gcm}, nil
}

// Encrypt encrypts plaintext. Returns nonce (12B) || ciphertext || tag (16B).
func (sc *SessionCipher) Encrypt(plaintext []byte) ([]byte, error) {
	nonce := make([]byte, NonceSize)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, fmt.Errorf("nonce generation: %w", err)
	}
	ct := sc.gcm.Seal(nil, nonce, plaintext, nil)
	sc.SendSeq++
	out := make([]byte, 0, NonceSize+len(ct))
	out = append(out, nonce...)
	out = append(out, ct...)
	return out, nil
}

// Decrypt decrypts a payload of nonce (12B) || ciphertext || tag (16B).
func (sc *SessionCipher) Decrypt(payload []byte) ([]byte, error) {
	if len(payload) < NonceSize+TagSize {
		return nil, errors.New("encrypted payload too short")
	}
	nonce := payload[:NonceSize]
	ct := payload[NonceSize:]
	plaintext, err := sc.gcm.Open(nil, nonce, ct, nil)
	if err != nil {
		return nil, errors.New("message integrity check failed — possible tampering or data corruption")
	}
	sc.RecvSeq++
	return plaintext, nil
}

// EncryptRaw performs one-shot AES-256-GCM encryption. Returns nonce || ct || tag.
func EncryptRaw(key, plaintext []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, errors.New("key must be 32 bytes")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	nonce := make([]byte, NonceSize)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}
	ct := gcm.Seal(nil, nonce, plaintext, nil)
	out := make([]byte, 0, NonceSize+len(ct))
	out = append(out, nonce...)
	out = append(out, ct...)
	return out, nil
}

// DecryptRaw performs one-shot AES-256-GCM decryption of nonce || ct || tag.
func DecryptRaw(key, payload []byte) ([]byte, error) {
	if len(key) != 32 {
		return nil, errors.New("key must be 32 bytes")
	}
	if len(payload) < NonceSize+TagSize {
		return nil, errors.New("encrypted payload too short")
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}
	nonce := payload[:NonceSize]
	ct := payload[NonceSize:]
	plaintext, err := gcm.Open(nil, nonce, ct, nil)
	if err != nil {
		return nil, errors.New("message integrity check failed — possible tampering or data corruption")
	}
	return plaintext, nil
}
